src/cells/base.mjs

```diff
@@ -5,91 +5,183 @@ import bigInt from 'big-integer';
 import * as help from '../help.mjs';
 import { Vector3vl } from '3vl';
 
+export const portGroupAttrs = {
+    'line.wire': {
+        stroke: '#4B4F6A',
+        x1: 0, y1: 0,
+        x2: undefined, y2: 0
+    },
+    'circle.port': {
+        magnet: undefined,
+        r: 7,
+        stroke: 'black',
+        fill: 'white',
+        strokeWidth: 2,
+        strokeOpacity: 0.5,
+        jointSelector: '.port'
+    },
+    'text.bits': {
+        ref: 'circle.port',
+        fill: 'black',
+        fontSize: '7pt'
+    },
+    'text.iolabel': {
+        yAlignment: 'middle',
+        fill: 'black',
+        fontSize: '8pt'
+    },
+    'path.decor': {
+        stroke: 'black',
+        fill: 'transparent',
+        d: undefined
+    }
+};
+
 // Common base class for gate models
 export const Gate = joint.shapes.basic.Generic.define('Gate', {
+    /* default properties */
+    propagation: 1,
+    label: '',
+    
     size: { width: 80, height: 30 },
     inputSignals: {},
     outputSignals: {},
-    propagation: 1,
     attrs: {
         '.': { magnet: false },
-        'rect.body': { width: 80, height: 30 },
-        'circle.port': { r: 7, stroke: 'black', fill: 'transparent', 'stroke-width': 2 },
-        'text.label': {
-            text: '', 'ref-x': 0.5, 'ref-dy': 2, 'text-anchor': 'middle', 
+        '.body': { stroke: 'black', strokeWidth: 2 },
+        'text': {
+            fontSize: '8pt',
             fill: 'black'
         },
-        'text.bits': {
-            fill: 'black'
+        'text.label': {
+            refX: .5, refDy: 3,
+            xAlignment: 'middle'
+        }
+    },
+    ports: {
+        groups: {
+            'in': {
+                position: 'left',
+                attrs: _.merge({}, portGroupAttrs, { 'line.wire': { x2: -20 }, 'circle.port': { magnet: 'passive', refX: -20 }, 'text.bits': { refDx: 3, refY: -4, textAnchor: 'start' }, 'text.iolabel': { refX: 5, textAnchor: 'start' } })
```

Comment:
This line is very long, the one four lines down is, too.

Reply:
True, I've split it down to several lines.